glean-core/rlb/src/net/mod.rs

```diff
@@ -115,11 +128,41 @@ impl UploadManager {
                             break;
                         }
                     }
+
+                    let status = inner.thread_running.load(Ordering::SeqCst);
+                    // asked to shut down. let's do it.
+                    if status == STATE_SHUTTING_DOWN {
+                        break;
+                    }
                 }
 
-                // Clear the running flag to signal that this thread is done.
-                inner.thread_running.store(false, Ordering::SeqCst);
+                // Clear the running flag to signal that this thread is done,
+                // but only if there's no shutdown thread.
+                let _ = inner.thread_running.compare_exchange(
+                    STATE_RUNNING,
+                    STATE_STOPPED,
+                    Ordering::SeqCst,
+                    Ordering::SeqCst,
+                );
             })
             .expect("Failed to spawn Glean's uploader thread");
+        *handle = Some(thread);
+    }
+
+    pub(crate) fn shutdown(&self) {
+        // mark as shutting down.
+        self.inner
+            .thread_running
+            .store(STATE_SHUTTING_DOWN, Ordering::SeqCst);
+
+        // take the thread handle out.
+        let mut handle = self.inner.handle.lock().unwrap();
+        let thread = handle.take();
+
+        if let Some(thread) = thread {
+            thread
+                .join()
```

Comment:
So this will work if our "we're killing it too early" is interrupting us in the middle of a `PingTask::Upload`, but if we're in a `PingTask::Wait`, then not only will we wait, we'll then not perform an upload attempt?

I'm tempted to put the `if SHUTTING_DOWN: break;` inside `PingTask::Upload` to make it more likely that we actually give an upload a try here. But then we might want to set the rate limiter to one that doesn't wait so we don't cycle from Wait to Wait for the full `3*interval` (3min?). Thoughts?

Reply:
With the current throttle we're not able to run into it 3 times though, because we do sleep the appropriate amount of time and so the next call after a sleep should be allowed.
Rust's `thread::sleep` handles spurious wakeups and will go back to sleep to reach the appropriate `Duration`.

Reply:
Thank you for moving the `if ShuttingDown: break` inside `Upload`. Even if we're stuck waiting, at least we'll give one upload a try after we're done.